// react-native-paper-tabs/lib/module/internal.native.js
import * as React from 'react';
import { Animated, Platform } from 'react-native';
export function useAnimatedText({
  childrenCount,
  tabIndex,
  position,
  offset,
  textColor,
  activeColor
}) {
  const childrenA = Array(childrenCount).fill(undefined);
  const positionWithOffset = Animated.add(position, offset);
  const inputRange = childrenA.map((_, i) => i);
  return {
    color: positionWithOffset.interpolate({
      inputRange: inputRange,
      outputRange: childrenA.map((_, i) => i === tabIndex ? activeColor : textColor)
    }),
    opacity: positionWithOffset.interpolate({
      inputRange: inputRange,
      outputRange: childrenA.map((_, i) => i === tabIndex ? 1 : 0.6)
    })
  };
}
export function useIndicator({
  childrenCount,
  position,
  offset,
  layouts,
  tabsLayout
}) {
  const [renderIndex, setRenderIndex] = React.useState(0);
  const style = React.useMemo(() => {
    /* eslint-disable @typescript-eslint/no-unused-vars  */
    // @ts-ignore
    let _ = renderIndex;
    const childrenA = Array(childrenCount).fill(undefined);
    const inputRange = childrenA.map((__, i) => i);
    const positionWithOffset = Animated.add(position, offset);
    return position && tabsLayout && layouts.current ? {
      transform: [{
        scaleX: positionWithOffset.interpolate({
          inputRange,
          outputRange: childrenA.map((__, i) => {
            var _layouts$current, _layouts$current$i;

            return ((_layouts$current = layouts.current) === null || _layouts$current === void 0 ? void 0 : (_layouts$current$i = _layouts$current[i]) === null || _layouts$current$i === void 0 ? void 0 : _layouts$current$i.width) || 0;
          })
        })
      }, {
        translateX: positionWithOffset.interpolate({
          inputRange,
          outputRange: childrenA.map((__, i) => {
            var _layouts$current2;

            const cl = (_layouts$current2 = layouts.current) === null || _layouts$current2 === void 0 ? void 0 : _layouts$current2[i];

            if (!cl) {
              return 0;
            }

            return (cl.x + cl.width / 2) / cl.width;
          })
        })
      }]
    } : null;
  }, [position, offset, tabsLayout, layouts, renderIndex, childrenCount]);
  const onUpdateTabLayout = React.useCallback(() => {
    setRenderIndex(prev => prev + 1);
  }, [setRenderIndex]);
  return [undefined, onUpdateTabLayout, style];
}
export function useOffsetScroller({
  index,
  offset,
  updateScroll,
  mode
}) {
  // we want native to scroll before the index changes
  const direction = React.useRef(undefined);
  React.useEffect(() => {
    // android does not work unfortunately
    if (offset && Platform.OS !== 'android' && mode === 'scrollable') {
      const id = offset.addListener(nOffset => {
        const newOffset = nOffset.value;
        const oldDirection = direction.current;

        if (newOffset > 0.1) {
          direction.current = 'next';
        } else if (newOffset < -0.1) {
          direction.current = 'prev';
        }

        if (direction.current) {
          if (oldDirection !== direction.current) {
            updateScroll(direction.current);
          }
        }
      });
      return () => {
        offset.removeListener(id);
      };
    }

    return undefined;
  }, [offset, updateScroll, direction, mode]);
  React.useEffect(() => {
    direction.current = undefined;
  }, [index]);
}
