// react-native-paper-tabs/lib/module/Swiper.native.js
import * as React from 'react';
import { View, Animated, Keyboard, StyleSheet } from 'react-native';
import ViewPager from 'react-native-pager-view';
import { TabsContext } from './context';
import TabsHeader from './TabsHeader';
const styles = StyleSheet.create({
  viewPager: {
    flex: 1
  }
});

function SwiperNative(props) {
  const {
    theme,
    dark,
    style,
    defaultIndex,
    iconPosition,
    showTextLabel,
    uppercase,
    mode,
    onChangeIndex,
    showLeadingSpace,
    disableSwipe
  } = props;
  const indexRef = React.useRef(defaultIndex || 0);
  const [index, setIndex] = React.useState(defaultIndex || 0);
  let children = props.children;
  const offset = React.useRef(new Animated.Value(0));
  const position = React.useRef(new Animated.Value(defaultIndex || 0));
  const isScrolling = React.useRef(false);
  const viewPager = React.useRef(undefined);
  React.useEffect(() => {
    if (index !== indexRef.current) {
      isScrolling.current = true;
      requestAnimationFrame(() => viewPager.current && viewPager.current.setPage(index));
    }

    indexRef.current = index;
    return undefined;
  }, [isScrolling, viewPager, index]);
  const onPageScrollStateChanged = React.useCallback(event => {
    Keyboard.dismiss();
    isScrolling.current = event.nativeEvent.pageScrollState !== 'idle';
  }, [isScrolling]);
  const onPageSelected = React.useCallback(e => {
    isScrolling.current = false;
    const i = e.nativeEvent.position;
    setIndex(i);
    onChangeIndex(i);
  }, [isScrolling, setIndex, onChangeIndex]);
  const goTo = React.useCallback(ind => {
    if (!isScrolling.current) {
      setIndex(ind);
    }
  }, [setIndex, isScrolling]);
  const renderProps = {
    index,
    goTo,
    children,
    theme,
    dark,
    style,
    position: position.current,
    offset: offset.current,
    iconPosition,
    showTextLabel,
    showLeadingSpace,
    uppercase,
    mode
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(TabsHeader, renderProps), /*#__PURE__*/React.createElement(TabsContext.Provider, {
    value: {
      goTo,
      index
    }
  }, /*#__PURE__*/React.createElement(ViewPager, {
    style: styles.viewPager,
    initialPage: index,
    scrollEnabled: !disableSwipe,
    onPageSelected: onPageSelected,
    ref: viewPager,
    onPageScrollStateChanged: onPageScrollStateChanged,
    onPageScroll: Animated.event([{
      nativeEvent: {
        position: position.current,
        offset: offset.current
      }
    }], {
      useNativeDriver: false
    })
  }, React.Children.map(children, (tab, tabIndex) => /*#__PURE__*/React.createElement(View, {
    style: styles.viewPager,
    key: tab.props.label || tabIndex
  }, tab)))));
}

export default SwiperNative;
