// react-native-paper-tabs/lib/module/Swiper.js
import * as React from 'react';
import { StyleSheet, View } from 'react-native';
import { TabsContext } from './context';
import TabsHeader from './TabsHeader';

function Swiper(props) {
  const {
    theme,
    dark,
    style,
    defaultIndex,
    onChangeIndex,
    iconPosition,
    showTextLabel,
    showLeadingSpace,
    uppercase,
    mode
  } = props;
  const [index, setIndex] = React.useState(defaultIndex || 0);
  const goTo = React.useCallback(ind => {
    setIndex(ind);
    onChangeIndex(ind);
  }, [setIndex, onChangeIndex]);
  let children = props.children;
  const currentScreen = children[index];

  if (!currentScreen) {
    return null;
  }

  const renderProps = {
    index,
    goTo,
    children,
    theme,
    dark,
    style,
    offset: undefined,
    position: undefined,
    iconPosition,
    showTextLabel,
    showLeadingSpace,
    uppercase,
    mode
  };
  return /*#__PURE__*/React.createElement(View, {
    style: styles.root
  }, /*#__PURE__*/React.createElement(TabsHeader, renderProps), /*#__PURE__*/React.createElement(TabsContext.Provider, {
    value: {
      goTo,
      index
    }
  }, currentScreen));
}

const styles = StyleSheet.create({
  root: {
    flex: 1
  }
});
export default Swiper;
